File: src/main/java/dev/zprestige/fire/module/misc/nointerpolation/NoInterpolation.java
package dev.zprestige.fire.module.misc.nointerpolation;

import dev.zprestige.fire.events.eventbus.annotation.RegisterListener;
import dev.zprestige.fire.module.Descriptor;
import dev.zprestige.fire.module.Module;
import dev.zprestige.fire.newbus.EventListener;
import dev.zprestige.fire.settings.impl.Slider;
import dev.zprestige.fire.settings.impl.Switch;
import net.minecraft.entity.player.EntityPlayer;

import java.util.ArrayList;
@Descriptor(description = "Removes interpolation between ticks")
public class NoInterpolation extends Module {
    public final Slider delay = Menu.Slider("Delay", 1.0f, 1.0f, 20.0f);
    public final Switch sneak = Menu.Switch("Sneak", false);
    public final Switch cancelAnimations = Menu.Switch("Cancel Animations", false);
    protected ArrayList<NoInterpolatedPlayer> noInterpolatedPlayers = new ArrayList<>();
    protected int i;

    public NoInterpolation(){
        eventListeners = new EventListener[]{
                new Frame3DListener(this),
                new TickListener(this)
        };
    }

    protected void fixPos(final EntityPlayer entityPlayer){
        entityPlayer.prevPosX = entityPlayer.posX;
        entityPlayer.prevPosY = entityPlayer.posY;
        entityPlayer.prevPosZ = entityPlayer.posZ;
        entityPlayer.prevChasingPosX = entityPlayer.posX;
        entityPlayer.prevChasingPosY = entityPlayer.posY;
        entityPlayer.prevChasingPosZ = entityPlayer.posZ;
    }

    protected void cancelAnimations(final EntityPlayer entityPlayer){
        entityPlayer.limbSwing = 0;
        entityPlayer.limbSwingAmount = 0;
        entityPlayer.prevLimbSwingAmount = 0;
        entityPlayer.rotationPitch = 0;
        entityPlayer.prevRotationPitch = 0;
    }
    
    protected boolean contains(final EntityPlayer entityPlayer){
        return noInterpolatedPlayers.stream().anyMatch(noInterpolatedPlayer -> noInterpolatedPlayer.getEntityPlayer().equals(entityPlayer));
    }

    protected static class NoInterpolatedPlayer {
        protected final EntityPlayer entityPlayer;
        protected double x, y, z;

        public NoInterpolatedPlayer(final EntityPlayer entityPlayer) {
            this.entityPlayer = entityPlayer;
            setPos();
        }

        public void updatePosition() {
            entityPlayer.setPosition(x, y, z);
        }

        public void setPos() {
            this.x = entityPlayer.posX;
            this.y = entityPlayer.posY;
            this.z = entityPlayer.posZ;
        }

        public EntityPlayer getEntityPlayer() {
            return entityPlayer;
        }

        public double getX() {
            return x;
        }

        public double getY() {
            return y;
        }

        public double getZ() {
            return z;
        }
    }
}

File: src/main/java/dev/zprestige/fire/module/misc/nointerpolation/NoInterpolation.kt
package dev.zprestige.fire.module.misc.nointerpolation

import dev.zprestige.fire.event.bus.EventListener
import dev.zprestige.fire.module.Descriptor
import dev.zprestige.fire.module.Module
import dev.zprestige.fire.settings.impl.Slider
import dev.zprestige.fire.settings.impl.Switch
import net.minecraft.entity.player.EntityPlayer

@Descriptor(description = "Removes interpolation between ticks")
class NoInterpolation : Module() {
    val delay: Slider = Menu.Slider("Delay", 1.0f, 1.0f, 20.0f)
    val sneak: Switch = Menu.Switch("Sneak", false)
    val cancelAnimations: Switch = Menu.Switch("Cancel Animations", false)
    var noInterpolatedPlayers = ArrayList<NoInterpolatedPlayer>()
    var i = 0

    init {
        eventListeners = arrayOf<EventListener<*, *>>(
            Frame3DListener(this),
            TickListener(this)
        )
    }

    fun fixPos(entityPlayer: EntityPlayer) {
        entityPlayer.prevPosX = entityPlayer.posX
        entityPlayer.prevPosY = entityPlayer.posY
        entityPlayer.prevPosZ = entityPlayer.posZ
        entityPlayer.prevChasingPosX = entityPlayer.posX
        entityPlayer.prevChasingPosY = entityPlayer.posY
        entityPlayer.prevChasingPosZ = entityPlayer.posZ
    }

    fun cancelAnimations(entityPlayer: EntityPlayer) {
        entityPlayer.limbSwing = 0f
        entityPlayer.limbSwingAmount = 0f
        entityPlayer.prevLimbSwingAmount = 0f
        entityPlayer.rotationPitch = 0f
        entityPlayer.prevRotationPitch = 0f
    }

    operator fun contains(entityPlayer: EntityPlayer): Boolean {
        return noInterpolatedPlayers.stream()
            .anyMatch { noInterpolatedPlayer: NoInterpolatedPlayer -> noInterpolatedPlayer.entityPlayer == entityPlayer }
    }

    class NoInterpolatedPlayer(val entityPlayer: EntityPlayer) {
        var x = 0.0
        var y = 0.0
        var z = 0.0

        init {
            setPos()
        }

        fun updatePosition() {
            entityPlayer.setPosition(x, y, z)
        }

        fun setPos() {
            x = entityPlayer.posX
            y = entityPlayer.posY
            z = entityPlayer.posZ
        }
    }
}